File: desktop/lib/core/utils/time_utils.dart
class TimeUtils {
  /// Group section title based on the note's last update date.
  static String getNoteSection(DateTime date) {
    final now = DateTime.now();
    final todayMidnight = DateTime(now.year, now.month, now.day);
    final yesterdayMidnight = todayMidnight.subtract(const Duration(days: 1));

    final itemDate = DateTime(date.year, date.month, date.day);

    if (itemDate.isAtSameMomentAs(todayMidnight)) {
      return 'Today';
    }

    if (itemDate.isAtSameMomentAs(yesterdayMidnight)) {
      return 'Yesterday';
    }

    // "This Week" (within the last 7 days, excluding today/yesterday)
    final sevenDaysAgo = todayMidnight.subtract(const Duration(days: 7));
    if (itemDate.isAfter(sevenDaysAgo)) {
      return 'This Week';
    }

    // "Last Month" (within the last 30 days, excluding the above)
    final thirtyDaysAgo = todayMidnight.subtract(const Duration(days: 30));
    if (itemDate.isAfter(thirtyDaysAgo)) {
      return 'Last Month';
    }

    // "Last Year" (within the last 365 days, excluding the above)
    final yearAgo = todayMidnight.subtract(const Duration(days: 365));
    if (itemDate.isAfter(yearAgo)) {
      return 'Last Year';
    }

    return 'Older';
  }

  /// Formats note time for display in the list.
  ///
  /// - Today: "3:45 PM"
  /// - Yesterday: "Yesterday"
  /// - Within last 7 days: Day of week (e.g., "Monday")
  /// - Older: Date format "MM/DD/YY" (e.g., "6/4/26")
  static String formatCardTime(DateTime date) {
    final now = DateTime.now();
    final todayMidnight = DateTime(now.year, now.month, now.day);
    final yesterdayMidnight = todayMidnight.subtract(const Duration(days: 1));
    final itemDate = DateTime(date.year, date.month, date.day);

    final hour =
        date.hour > 12 ? date.hour - 12 : (date.hour == 0 ? 12 : date.hour);
    final ampm = date.hour >= 12 ? 'PM' : 'AM';
    final minute = date.minute.toString().padLeft(2, '0');
    final timeStr = '$hour:$minute $ampm';

    if (itemDate.isAtSameMomentAs(todayMidnight)) {
      return timeStr;
    }

    if (itemDate.isAtSameMomentAs(yesterdayMidnight)) {
      return 'Yesterday';
    }

    final sevenDaysAgo = todayMidnight.subtract(const Duration(days: 7));
    if (itemDate.isAfter(sevenDaysAgo)) {
      const weekdays = [
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
        'Sunday',
      ];
      return weekdays[date.weekday - 1];
    }

    final yearStr = date.year.toString();
    final shortYear = yearStr.length >= 4 ? yearStr.substring(2) : yearStr;
    return '${date.month}/${date.day}/$shortYear';
  }

  /// Formats a date for the editor header: "4 Jun 2026 at 3:45 PM".
  static String formatEditorHeader(DateTime date) {
    const months = [
      'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
      'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec',
    ];
    final hour =
        date.hour > 12 ? date.hour - 12 : (date.hour == 0 ? 12 : date.hour);
    final ampm = date.hour >= 12 ? 'PM' : 'AM';
    final minute = date.minute.toString().padLeft(2, '0');
    return '${date.day} ${months[date.month - 1]} ${date.year} at $hour:$minute $ampm';
  }
}

File: note-desktop/Sources/Utils/TimeUtils.swift
import Foundation

public final class TimeUtils {
    nonisolated(unsafe) private static let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
    
    public static func stringFromDate(_ date: Date) -> String {
        return isoFormatter.string(from: date)
    }
    
    public static func dateFromString(_ string: String) -> Date? {
        if let date = isoFormatter.date(from: string) {
            return date
        }
        let basicFormatter = ISO8601DateFormatter()
        return basicFormatter.date(from: string)
    }
    
    /// Group section title based on the creation date of the note.
    public static func getNoteSection(for date: Date) -> String {
        let calendar = Calendar.current
        let now = Date()
        
        let todayMidnight = calendar.startOfDay(for: now)
        guard let yesterdayMidnight = calendar.date(byAdding: .day, value: -1, to: todayMidnight) else { return "Older" }
        
        let itemDate = calendar.startOfDay(for: date)
        
        if itemDate == todayMidnight {
            return "Today"
        }
        
        if itemDate == yesterdayMidnight {
            return "Yesterday"
        }
        
        // "Previous 7 Days"
        guard let sevenDaysAgo = calendar.date(byAdding: .day, value: -7, to: todayMidnight) else { return "Older" }
        if itemDate >= sevenDaysAgo {
            return "Previous 7 Days"
        }
        
        // "Previous 30 Days"
        guard let thirtyDaysAgo = calendar.date(byAdding: .day, value: -30, to: todayMidnight) else { return "Older" }
        if itemDate >= thirtyDaysAgo {
            return "Previous 30 Days"
        }
        
        // Same Year check
        let nowYear = calendar.component(.year, from: now)
        let itemYear = calendar.component(.year, from: date)
        
        if itemYear == nowYear {
            let monthFormatter = DateFormatter()
            monthFormatter.dateFormat = "MMMM"
            return monthFormatter.string(from: date)
        }
        
        return "\(itemYear)"
    }
    
    /// Formats note creation/update time for display in the list card.
    public static func formatCardTime(for date: Date) -> String {
        let calendar = Calendar.current
        let now = Date()
        
        let todayMidnight = calendar.startOfDay(for: now)
        guard let yesterdayMidnight = calendar.date(byAdding: .day, value: -1, to: todayMidnight) else { return "" }
        let itemDate = calendar.startOfDay(for: date)
        
        let timeFormatter = DateFormatter()
        timeFormatter.dateFormat = "h:mm a"
        let timeStr = timeFormatter.string(from: date)
        
        if itemDate == todayMidnight {
            return timeStr
        }
        
        if itemDate == yesterdayMidnight {
            return "Yesterday"
        }
        
        guard let sevenDaysAgo = calendar.date(byAdding: .day, value: -7, to: todayMidnight) else { return "" }
        if itemDate > sevenDaysAgo {
            let dayFormatter = DateFormatter()
            dayFormatter.dateFormat = "EEEE"
            return dayFormatter.string(from: date)
        }
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "M/d/yy"
        return dateFormatter.string(from: date)
    }
    
    /// Formats a date for the editor header: "4 Jun 2026 at 3:45 PM".
    public static func formatEditorHeader(for date: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "d MMM yyyy 'at' h:mm a"
        return formatter.string(from: date)
    }
}
